### PR TITLE
add the ability to add more idf components to an existing setup

### PR DESCRIPTION
**Pull request by nielsnl68**
# What does this implement/fix?

When a one component added some extra esp_idf components a second attempt was ignored.
With this PR the new components and submodules are bying added to the earlier request.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/esp32/__init__.py [hunk 1]
```diff
@@ -169,9 +169,28 @@ def add_idf_component(
             KEY_REF: ref,
             KEY_PATH: path,
             KEY_REFRESH: refresh,
-            KEY_COMPONENTS: components,
-            KEY_SUBMODULES: submodules,
+            KEY_COMPONENTS: components if components is not None else [],
```
Comment (jesserockz) [inline esphome/components/esp32/__init__.py]: Line 164-165 above ensures it is at least an empty list already
```suggestion
            KEY_COMPONENTS: components,
```

File: esphome/components/esp32/__init__.py [hunk 2]
```diff
@@ -169,9 +169,28 @@ def add_idf_component(
...
+            KEY_SUBMODULES: submodules if submodules is not None else [],
```
Comment (jesserockz) [inline esphome/components/esp32/__init__.py]: This will break the `clone_or_update` function that checks explicitly for `None`

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/esp32/__init__.py [hunk 3]
```diff
@@ -172,6 +172,24 @@ def add_idf_component(
             KEY_COMPONENTS: components,
             KEY_SUBMODULES: submodules,
         }
+    else:
+        if components is not None:
```
Comment (jesserockz) [inline esphome/components/esp32/__init__.py:177]: Lets pull this out to clean it up a bit
```suggestion
        component_config = CORE.data[KEY_ESP32][KEY_COMPONENTS][name]
        if components is not None:
```

File: esphome/components/esp32/__init__.py [hunk 4]
```diff
@@ -172,6 +172,19 @@ def add_idf_component(
...
+            component_config[KEY_COMPONENTS] = list(
+                set(component_config[KEY_COMPONENTS] + components)
+            )
+        if submodules is not None:
+            if component_config[KEY_SUBMODULES] is None:
+                component_config[KEY_SUBMODULES] = submodules
+            else:
+                component_config[KEY_SUBMODULES] = list(
+                    set([KEY_SUBMODULES] + submodules)
```
Comment (jesserockz) [inline esphome/components/esp32/__init__.py]: Missing a critical piece :wink: 
```suggestion
                    set(component_config[KEY_SUBMODULES] + submodules)
```

Comment (nielsnl68) [reply esphome/components/esp32/__init__.py]: I think i was sleeping at that point :(

**Review by jesserockz — APPROVED** [review]

**Pull request merged**